### DISCUSSION TITLE
How to parse an entire project?

### DISCUSSION THREAD

**Discussion by MilesConn** [swc-project/swc#5787] [Q&A]
I'm working on a simple transpiler to extract IDs from `.tsx` files and put them into another format. I have been using the visitor trait to `vist_lit` and `visit_tpl`. What I'm wondering is how do I parse an entire project given the root folder? I was hoping not to have to implement this logic myself. Any tips? 

I tried searching the `swc` codebase but couldn't find anything immediately obvious. Initially I was thinking I could start at find all `tsx` files and collect them into a vec and parse each one. However, I noticed that my `next.js` projects don't have the same structure ie everything in `src`.

**Comment by pictographer** [comment]
Trying to do something similar. Looking to observe function definitions and calls, not to transpile the code, but to make test templates.

Haven't found a way of observing file open/close events.

**Reply by MilesConn** [reply]
I have two possible solutions. One is I wrote this code to kinda sorta collect all the relevant files. It's not very smart but it gets the job done 
```rust
pub struct FindFileOptions {
    extension_to_look_for: Vec<OsString>,
    directories_to_ignore: Vec<OsString>,
}

impl Default for FindFileOptions {
    fn default() -> Self {
        Self {
            extension_to_look_for: vec!["tsx".into()],
            directories_to_ignore: vec!["node_modules".into()],
        }
    }
}

impl FindFileOptions {
    pub fn is_ignored_file(&self, path: &PathBuf) -> bool {
        if path.is_dir() {
            let file_name = path.file_name().unwrap();
            return self.directories_to_ignore.contains(&file_name.into());
        } else if path.is_file() {
            if let Some(ext) = path.extension() {
                return !self.extension_to_look_for.contains(&ext.into());
            }
        }
        true
    }
}

pub fn find_files(base_path: &PathBuf, find_files_options: &FindFileOptions) -> Vec<PathBuf> {
    let mut collectedFiles = Vec::new();

    for file in base_path.read_dir().unwrap() {
        let path = file.unwrap().path();
        if path.is_dir() && !find_files_options.is_ignored_file(&path) {
            collectedFiles.append(&mut find_files(&path, find_files_options));
        } else if path.is_file() && !find_files_options.is_ignored_file(&path) {
            collectedFiles.push(path);
        }
    }

    collectedFiles
}
```

the other thing I'd recommend is the [new rust typechecker](https://github.com/dudykr/stc). I only glanced over the project but it should have some easier to grok examples of how to correctly read an entire project. Hope that helps
